functorch/csrc/BatchRulesViews.cpp

```diff
@@ -342,6 +342,19 @@ std::tuple<Tensor,optional<int64_t>> slice_backward_batch_rule(
   return std::make_tuple(std::move(result), 0);
 }
 
+std::tuple<Tensor, optional<int64_t>> view_batching_rule(const Tensor& self, optional<int64_t> self_bdim, IntArrayRef size) {
+  auto self_ = moveBatchDimToFront(self, self_bdim);
```

Comment:
I think if you move to front it wont be a view

Reply:
@vfdev-5 Thanks for looking at it.
I thought the bdim are mostly hidden from users so it is ok to shuffle around (which should keep it as a view)

Since I was porting it, I thought the tests would cover the cases. How did you figure that #137 was wrong? #156 does not mention what was wrong about that implementation.

Reply:
I think it does return view

```python
import torch
import functorch

t = torch.randn(3, 9, 3)

def fn(x, sizes):
    return x.view(sizes)

def assert_is_view(t1, t2):
    assert t1.storage().data_ptr() == t2.storage().data_ptr()

assert_is_view(functorch.vmap(fn, in_dims=(0, None))(t, (27, 1)), t)
assert_is_view(functorch.vmap(fn, in_dims=(1, None))(t, (1, 1, 3, 3)), t)
assert_is_view(functorch.vmap(fn, in_dims=(2, None))(t, (1, 3, 9)), t)

```

Reply:
Well, in my PR I removed few tests that should be present: https://github.com/facebookresearch/functorch/pull/137/files#diff-ee63347c8ea255a43be0bd8e3c7b46c025877f06f3ab22538e409a5c4daac678L2436-L2437

Yes, seems like movedim (inside moveBatchDimToFront) returns a view but non-contiguous... I do not know if it is expected by user...

I think what we need to do is to insert bdim where it should be appropriate. For example, input tensor is [d1, d2, ... bdim,  dk, ... dN] and view shape is [D1, ... DN]. We have to insert bdim where it is appropriate without tweaking strides

Reply:
I am not sure if there is one correct output 🤔 

For Input Shape (9, B0, 3) -> View Shape (3, 1, 1, 3, 3), are we expecting (3, 1, 1, 3, B0, 3)? I think as long as we track the Batch dim correctly (it should be ok).

@zou3519 @Chillee what is your opinion on the same?

Reply:
I think we can figure out how to keep the stride for bdim instead of moving it somewhere